### DISCUSSION TITLE
Cannot Have Nullable Return of Async Generic Method

### DISCUSSION THREAD

**Discussion by kalaskarsanket** [dotnet/csharplang#4231] [Q&A]
---
**Issue moved from microsoft/dotnet#1249**
- Please respond to @mysteryx93.

---

_From @mysteryx93 on Thursday, September 10, 2020 8:17:31 PM_

With C# 8 Nullables, you can define a nullable generic like this

    [return: MaybeNull]
    public T Get<T>() => default;

However, I'm running into a problem with a Generic Async method.

    [return: MaybeNull]
    public async Task<T> GetAsync<T>() => default;

This would tell that Task is nullable, not T.

The only solution for now would be to add ! after default, but then the IDE will wrongly indicate to the users that the type is not null when it can be. This can work for private use (and can easily lead to errors), but not for a public API.

Is there any other solution?

Actually what I'd really need is a generic constraint "where T : nullable" because I need to document to pass in nullable types.

Currently, the only way of doing so is to have 2 copies, one with "where T : class" and one with "where T : struct", and using T? where ? is interpreted differently in each copy (Nullable<T> vs nullable attribute). That can lead to doubling a lot of code.

**Comment by kalaskarsanket** [comment]
---
**Issue moved from microsoft/dotnet#1249**
- Please respond to @mysteryx93.

---

_From @mysteryx93 on Saturday, September 19, 2020 5:51:30 PM_

Since there has been no response, here's my take on this.

The .NET team could fix it in 2 ways.

1. Add a "nullable" constraint

2. Tweaking MaybeNull attribute to work in the above case


_On a different topic_, a 3rd suggestion to help working with Generics

    if (typeof(T) == typeof(JsonElement))
    {
        return (T)(object)doc.RootElement;
    }

This works but has unnecessary boxing/unboxing, and disables type-checking

The compiler should recognize this as valid IMO

    if (typeof(T) == typeof(JsonElement))
    {
        return (T)doc.RootElement;
    }

**Comment by 333fred** [comment]
In C# 9, we've enabled unconstrained `T?`, so you can simply write `Task<T?>`. Note that this still behaves like regular generics: `int` does not become `int?` here, but rather just becomes `int`. If the type you substitute is a reference type, then the return will be nullable.

**Reply by zzzachzzz** [reply]
Really? I have this specified in my csproj file:
```
<TargetFramework>netcoreapp3.1</TargetFramework>
<LangVersion>9.0</LangVersion>
<Nullable>enable</Nullable>
```
And yet this method: `public async Task<T?> Get<T>(CacheEntry<T> cacheEntry)`  
is yielding the error:
```
error CS8627: A nullable type parameter must be known to be a value type or non-nullable reference type. Consider adding a 'class', 'struct', or type constraint.
```

**Reply by cston** [reply]
@zzzachzzz, it looks like that error message may be reported from an earlier version of the compiler.

In C#9, the error message for CS8627 was updated to mention the language version:

"error CS8627: A nullable type parameter must be known to be a value type or non-nullable reference type **unless language version '9.0' or greater is used**. Consider **changing the language version or** adding a 'class', 'struct', or type constraint."

With C#9, I was able to compile a method with the signature in your example.

**Reply by zzzachzzz** [reply]
@cston Meaning that netcoreapp3.1 may not be able to leverage all C# 9 features, even though it can be specified via `<LangVersion>`? That's my takeaway from this excerpt of the [language versioning docs](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/configure-language-version):
>C# 10 is supported only on .NET 6 and newer versions. C# 9 is supported only on .NET 5 and newer versions. C# 8.0 is supported only on .NET Core 3.x and newer versions.

**Reply by jnm2** [reply]
I'm using C# 10 with .NET Framework 4.8. Maybe what you're missing here is `where T : default`. I thought that only mattered for overrides though.

**Reply by RikkiGibson** [reply]
> I'm using C# 10 with .NET Framework 4.8. Maybe what you're missing here is `where T : default`. I thought that only mattered for overrides though.

We haven't introduced an actionable diagnostic for this yet? 😕

**Reply by zzzachzzz** [reply]
I was going to test with `<LangVersion>` set to `10.0` but it's not available for me:
```
$ dotnet /usr/share/dotnet/sdk/3.1.417/Roslyn/bincore/csc.dll /langversion:\?
Supported language versions:
default
1
2
3
4
5
6
7.0
7.1
7.2
7.3
8.0 (default)
9.0
latestmajor
preview
latest
```

**Reply by zzzachzzz** [reply]
When changing the method signature to the following:
`async Task<T?> ICache.Get<T>(CacheEntry<T> cacheEntry) where T : default`

Lines where I `return null` still yield:
```
Cannot convert null to type parameter 'T' because it could be a non-nullable value type. Consider using 'default(T)' instead.
```

**Reply by jnm2** [reply]
@zzzachzzz You need to install the .NET 6 SDK to compile C# 10. That's not the same thing as needing your app to run on .NET 6; the .NET 6 SDK can build .NET Core 3.1 and even .NET Framework apps.

**Reply by jnm2** [reply]
@zzzachzzz You simply can't convert `null` to `T` in any version of the C# language. Unless you have `where T : struct` or `where T : class`, `T?` means "defaultable" rather than "nullable" which is what @333fred was saying above.

If `T` appears in the parameter section of a method signature, you can define overloads side by side with `where T : struct` and `where T : class`. I do this sometimes.
